### PR TITLE
Create ClosureKeyPresses.wurst

### PR DESCRIPTION
**Pull request by WareditorGit**
Closures for the new key presses natives.

### CODE REVIEW

File: wurst/closures/ClosureKeyPresses.wurst [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+        listener = listener.next 
+
+
+/* KEY PRESSES LISTENER */
+
+abstract class KeyPressListener
+    oskeytype key 
+    int metaKey
+    bool onKeyDown
+    thistype next = null
```
Comment (Frotty) [inline wurst/closures/ClosureKeyPresses.wurst]: we should probably make a module for this now that we use it at multiple occasions.
Basically like LinkedListModule but with custom add/get/remove etc.
Not in this PR tho ofc, just noting.

File: wurst/closures/ClosureKeyPresses.wurst [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+        let id = key.getHandleId() 
+        let first = KEY_MAP.get(id)
+        if first == this 
+            if next == null 
+                KEY_MAP.remove(id)
+            else 
+                KEY_MAP.put(id, next)
+        else if prev != null
+            prev.next = next
+        next.prev = prev
```
Comment (Frotty) [inline wurst/closures/ClosureKeyPresses.wurst]: missing eof newline - editorconfig should take care of this unless you use github web to submit new file PR.

File: wurst/closures/ClosureKeyPresses.wurst [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+
+/*  
+    Example usage:
+        onKeyPress(OSKEY_A) -> 
+            print("A player pressed on the A key.")
+
+        onKeyPress(OSKEY_B, OSKEY_META.SHIFT) -> 
+            print("A player pressed on the B key while holding a SHIFT key.")
+
+        onKeyPress(OSKEY_C, false) ->
```
Comment (Frotty) [inline wurst/closures/ClosureKeyPresses.wurst]: Seems a little weird, why not `onKeyRelease` ?

File: wurst/closures/ClosureKeyPresses.wurst [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+            print("A player pressed on the B key while holding a SHIFT key.")
+
+        onKeyPress(OSKEY_C, false) -> 
+            print("A player released the C key.")
+
+        onKeyPress(players[0], OSKEY_D, OSKEY_META.ALT, false) -> 
+            print("Player 0 released the D key while holding an ALT key.")
+
+    Important:
+        * When the metakey NONE is passed or when no metaKey is given like in `onKeyPress(OSKEY_A, () -> print("A"))`,
```
Comment (Frotty) [inline wurst/closures/ClosureKeyPresses.wurst]: Okay, as discussed in chat, I think as part of a combination that should be the correct behavior,
but this leaves me 2 questions, does the R-release event fire if I:
1. press R
2. press meta key
3. release R

and 
1. press R
2. press meta key
3. release meta key
4. release R

?

File: wurst/closures/ClosureKeyPresses.wurst (hunk 3 above)
Comment (WareditorGit) [reply wurst/closures/ClosureKeyPresses.wurst]: good idea

File: wurst/closures/ClosureKeyPresses.wurst (hunk 4 above)
Comment (WareditorGit) [reply wurst/closures/ClosureKeyPresses.wurst]: In the first case it doesn't fire.
In the second it does.

Comment (Frotty) [reply wurst/closures/ClosureKeyPresses.wurst]: I guess that's fine?

**Comment by Cokemonkey11** [comment]
Looks pretty good overall. A few notes:

- code comments are lacking
- package documentation explains how to use it nicely, but there's no blurb for what this package is, why it exists, and what it's for
- [opinion] pure interfaces are a bit nicer than abstract classes

**Comment by WareditorGit** [comment]
@Cokemonkey11 
* What comments you feel should be added?
* I will add an intro phrase
* I feel it would be too messy to extract all the members of the class to just have _pure_ interface

**Comment by Cokemonkey11** [comment]
> What comments you feel should be added?

Comments should describe "why", clarify anything that wouldn't be obvious to you when looking at your own code after 2+ years

> I feel it would be too messy to extract all the members of the class to just have pure interface

Dunno, I didn't think about it that deeply, but usually it can be done

**Comment by WareditorGit** [comment]
This package seems simple enough to be maintained at any point in the future without any added comments. That's why I asked you to point out some specific point I should comment.

`it can be done`
It can be done sure, but I don't feel it will make the package better.

**Comment by Cokemonkey11** [comment]
> This package seems simple enough to be maintained at any point in the future without any added comments. That's why I asked you to point out some specific point I should comment.

No you didn't

> It can be done sure, but I don't feel it will make the package better.

Pure interfaces make everything better

File: wurst/closures/ClosureKeyPresses.wurst [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+    return onKeyPress(key, OSKEY_META.NONE, true, listener)
+
+public function onKeyPress(oskeytype key, OSKEY_META metaKey, KeyPressListener listener) returns KeyPressListener
+    return onKeyPress(key, metaKey, true, listener)
+
+public function onKeyRelease(oskeytype key, KeyPressListener listener) returns KeyPressListener
+    return onKeyPress(key, OSKEY_META.NONE, false, listener)
+
+public function onKeyPress(oskeytype key, OSKEY_META metaKey, bool onKeyDown, KeyPressListener listener) returns KeyPressListener
+    for i=0 to bj_MAX_PLAYERS
```
Comment (Frotty) [inline wurst/closures/ClosureKeyPresses.wurst]: i=0 -> i = 0

**Comment by Frotty** [comment]
Abstract class seems fine, but as I mentioned we should pull out the "list" logic as module at some point.
Looks good to me now.

**Comment by WareditorGit** [comment]
Created the list module and made a PR for it #268
Once everything is merged, I will make another PR to make the new packages use the module.

**Comment by Frotty** [comment]
Sry, was a bit busy. Merging now.

>  I will make another PR to make the new packages use the module.

Thanks a lot 👍

**Pull request merged**